{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a14a676e-84e1-4705-8c0a-8c73f8a7b0b5",
   "metadata": {},
   "source": [
    "# Results on the full dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c241c1a-9e1d-432e-a519-11efca9aa7f6",
   "metadata": {},
   "source": [
    "The first result still starts from an overview of the MSA files, which we want to inspect **the gap columns** and **identity columns**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dd2a9bf-80db-43f9-b06e-3d444c39b39d",
   "metadata": {},
   "source": [
    "Before that, let's download the results that I precomputed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cb0d9d5b-7a56-4cc6-a5af-ed7426589d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/CoBi-course"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "359d011c-1a9a-4322-9e41-04fb81ba03a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Already up to date.\n"
     ]
    }
   ],
   "source": [
    "git pull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8967a94-5799-48f0-916a-f32d7bed70d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "source ~/env_func/bin/activate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2b0e73d8-5f68-457a-92a2-b614f4411251",
   "metadata": {},
   "outputs": [],
   "source": [
    "python ~/CoBi-course/func/multifunction.py summarizealns ~/CoBi-course/full_dataset_msa/*.fas -o ~/CoBi-course/full_dataset_concat/msa.full.summary.svg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20083ca8-e86b-4e12-8028-157440c0a37d",
   "metadata": {},
   "source": [
    "> Now double click on the `msa.full.summary.svg` file on the left side or open a `python` Console and type in the following commands and the `shift+enter` key to see the figure:\n",
    "> \n",
    "> ```python\n",
    "from IPython.display import SVG\n",
    "from pathlib import Path\n",
    "file_path = \"/home/jovyan/CoBi-course/full_dataset_concat/msa.full.summary.svg\"\n",
    "display(SVG(Path(file_path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c9164ff2-ec2d-4c7b-a78c-2304047622af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        ____________ Anthoceros_agrestis_Bonn\n",
      "                      ,|\n",
      "                      || __________ Marchantia_polymorpha\n",
      "                      |||\n",
      "                      | |________ Takakia_lepidozioides\n",
      "  ____________________|\n",
      " |                    |   _____________ Adiantum_capillus-veneris\n",
      " |                    | _|\n",
      " |                    || |     ____________ Amborella_trichopoda\n",
      "_|                    || |____|\n",
      " |                     |      |______ Cycas_panzhihuaensis\n",
      " |                     |\n",
      " |                     |______________ Selaginella_moellendorffii\n",
      " |\n",
      " | Zygnema_circumcarinatum_SAG_698-1b\n",
      "\n"
     ]
    }
   ],
   "source": [
    "python ~/CoBi-course/func/multifunction.py reroot ~/CoBi-course/full_dataset_concat/concatenated.msa.treefile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "562dded4-6a62-4034-8ac2-7f95a4bf29e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/CoBi-course/func/multifunction.py:162: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  fig.tight_layout()\n"
     ]
    }
   ],
   "source": [
    "python ~/CoBi-course/func/multifunction.py ratedis ~/CoBi-course/full_dataset_concat/mlerate_concat.mlrate -o ~/CoBi-course/full_dataset_concat/mlerate_concat.mlrate.hist.svg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d6b89ff-2736-4472-9f08-6f2f153a8859",
   "metadata": {},
   "source": [
    "> Double click on the `msa.full.summary.svg` file on the left side or open a `python` Console and type in the following commands and the `shift+enter` key to see the figure:\n",
    "> \n",
    "> ```python\n",
    "from IPython.display import SVG\n",
    "from pathlib import Path\n",
    "file_path = \"/home/jovyan/CoBi-course/full_dataset_concat/mlerate_concat.mlrate.hist.svg\"\n",
    "display(SVG(Path(file_path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b8237400-681c-47c3-a1e3-b13822300591",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Concordance factor statistics\n",
      "# This file can be read in MS Excel or in R with command:\n",
      "#   tab=read.table('sCF_concat.cf.stat',header=TRUE)\n",
      "# Columns are tab-separated with following meaning:\n",
      "#   ID: Branch ID\n",
      "#   sCF: Site concordance factor averaged over 100 quartets (=sCF_N/sN %)\n",
      "#   sCF_N: sCF in absolute number of sites\n",
      "#   sDF1: Site discordance factor for alternative quartet 1 (=sDF1_N/sN %)\n",
      "#   sDF1_N: sDF1 in absolute number of sites\n",
      "#   sDF2: Site discordance factor for alternative quartet 2 (=sDF2_N/sN %)\n",
      "#   sDF2_N: sDF2 in absolute number of sites\n",
      "#   sN: Number of informative sites averaged over 100 quartets\n",
      "#   Label: Existing branch label\n",
      "#   Length: Branch length\n",
      "ID\tsCF\tsCF_N\tsDF1\tsDF1_N\tsDF2\tsDF2_N\tsN\tLabel\tLength\n",
      "9\t30.58\t202.59\t37.72\t249.87\t31.7\t209.98\t662.44\t\t0.0296968\n",
      "10\t37.48\t665.62\t30.24\t537.08\t32.28\t573.2\t1775.9\t\t0.0400243\n",
      "11\t33.51\t265.91\t29.01\t230.21\t37.49\t297.48\t793.6\t\t0.0332873\n",
      "12\t35.74\t225.65\t33.94\t214.45\t30.32\t191.41\t631.51\t\t0.072917\n",
      "13\t57.54\t1175.74\t21.46\t438.51\t21.01\t429.32\t2043.57\t\t0.170028\n"
     ]
    }
   ],
   "source": [
    "cat ~/CoBi-course/full_dataset_concat/sCF_concat.cf.stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "336d32f2-4db8-42e2-bd9c-e910f6d5d5b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Concordance factor statistics\n",
      "# This file can be read in MS Excel or in R with command:\n",
      "#   tab=read.table('gCF_concat.cf.stat',header=TRUE)\n",
      "# Columns are tab-separated with following meaning:\n",
      "#   ID: Branch ID\n",
      "#   gCF: Gene concordance factor (=gCF_N/gN %)\n",
      "#   gCF_N: Number of trees concordant with the branch\n",
      "#   gDF1: Gene discordance factor for NNI-1 branch (=gDF1_N/gN %)\n",
      "#   gDF1_N: Number of trees concordant with NNI-1 branch\n",
      "#   gDF2: Gene discordance factor for NNI-2 branch (=gDF2_N/gN %)\n",
      "#   gDF2_N: Number of trees concordant with NNI-2 branch\n",
      "#   gDFP: Gene discordance factor due to polyphyly (=gDFP_N/gN %)\n",
      "#   gDFP_N: Number of trees decisive but discordant due to polyphyly\n",
      "#   gN: Number of trees decisive for the branch\n",
      "#   Label: Existing branch label\n",
      "#   Length: Branch length\n",
      "# *NOTE*: (gCF+gDF1+gDF2+gDFP) = 100% and (gCF_N+gDF1_N+gDF2_N+gDFP_N) = gN\n",
      "ID\tgCF\tgCF_N\tgDF1\tgDF1_N\tgDF2\tgDF2_N\tgDFP\tgDFP_N\tgN\tLabel\tLength\n",
      "9\t24\t24\t19\t19\t14\t14\t43\t43\t100\t\t0.0297025\n",
      "10\t43\t43\t15\t15\t9\t9\t33\t33\t100\t\t0.0400332\n",
      "11\t27\t27\t8\t8\t17\t17\t48\t48\t100\t\t0.0332955\n",
      "12\t48\t48\t18\t18\t10\t10\t24\t24\t100\t\t0.0729307\n",
      "13\t94\t94\t1\t1\t1\t1\t4\t4\t100\t\t0.17006\n"
     ]
    }
   ],
   "source": [
    "cat ~/CoBi-course/full_dataset_concat/gCF_concat.cf.stat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b59bd1df-1f23-47f1-93e7-fe83251b7829",
   "metadata": {},
   "source": [
    "If you want to know which id corresponds to which branch, print out the `gCF_concat.cf.branch` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f5adcc88-7a65-4f61-86b6-0978ad105a16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(Zygnema_circumcarinatum_SAG_698-1b:0.7198189295,(Anthoceros_agrestis_Bonn:0.4258604249,(Marchantia_polymorpha:0.3466936537,Takakia_lepidozioides:0.2919156572)10:0.0400331738)9:0.0297024815,((Adiantum_capillus-veneris:0.4661878454,(Amborella_trichopoda:0.4010445009,Cycas_panzhihuaensis:0.2139447855)13:0.1700597607)12:0.0729306986,Selaginella_moellendorffii:0.5047256117)11:0.0332955042)8;\n"
     ]
    }
   ],
   "source": [
    "cat ~/CoBi-course/full_dataset_concat/gCF_concat.cf.branch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cc897b27-1754-460c-8239-2cc8b46669d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "python ~/CoBi-course/func/multifunction.py scfgcfconstrast ~/CoBi-course/full_dataset_concat/sCF_concat.cf.stat  ~/CoBi-course/full_dataset_concat/gCF_concat.cf.stat -o ~/CoBi-course/full_dataset_concat/sCF_gCF.comparison.svg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba100e82-ba48-493a-9551-d88fec58cb66",
   "metadata": {},
   "source": [
    "> Again, double click on the `sCF_gCF.comparison.svg` file on the left side or open a `python` Console and type in the following commands and the `shift+enter` key to see the figure:\n",
    "> \n",
    "> ```python\n",
    "from IPython.display import SVG\n",
    "from pathlib import Path\n",
    "file_path = \"/home/jovyan/CoBi-course/full_dataset_concat/sCF_gCF.comparison.svg\"\n",
    "display(SVG(Path(file_path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7aaa0b3c-7b8f-4216-9579-f8a3591c52b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The PCC and P value between sCF and gCF: 0.980062689217865, 0.0033692299443270637\n",
      "The SCC and P value between sCF and gCF: 0.8999999999999998, 0.03738607346849874\n"
     ]
    }
   ],
   "source": [
    "python ~/CoBi-course/func/multifunction.py correlationtest ~/CoBi-course/full_dataset_concat/sCF_concat.cf.stat  ~/CoBi-course/full_dataset_concat/gCF_concat.cf.stat"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb1b9f8-b9d5-4e8b-9ba1-55fcceb006d6",
   "metadata": {},
   "source": [
    "The result from `phytop` has been precomputed and at `~/CoBi-course/full_dataset_concat/astral.result.pdf`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b2eb6fb-10fd-483e-875d-f4ab3057bf7d",
   "metadata": {},
   "source": [
    "# MSA simulation and quantify gene tree estimation error (GTEE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8615cb5-99eb-48b4-ae41-3cfd7e5b6568",
   "metadata": {},
   "source": [
    "Here we follow the analyses done in [**Cai et al. (2021)**](https://doi.org/10.1093/sysbio/syaa083) to quantify the GTEE in our dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "211c2139-9313-4374-bced-4cad47897bfc",
   "metadata": {},
   "source": [
    "Step 1: simulate MSA based on your original MSA and best-fit model. The command line is as below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c942b5-5d1b-4529-b787-de3e996b2130",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir CoBi-GTEE;cd CoBi-GTEE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f3faa44-dc98-4e10-99ef-aa91c70bb05f",
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree3 -s ~/CoBi-course/full_dataset_concat/concatenated.msa -m LG+F+R4 --alisim concatenated.msa.alignment_mimic --out-format fasta -t ~/CoBi-course/full_dataset_concat/concatenated.msa.treefile --num-alignments 100 -pre concatenated.msa.seqsimu -T 6"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "039d20cb-7da8-4824-b32d-68439aceaf21",
   "metadata": {},
   "source": [
    "Step 2: calculate sCF for each simulated MSA. Below I give one example. You can submit jobs to `HPC` to finish the entire 100 MSAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1250f163-7b76-48c7-a649-daf7f2b9b351",
   "metadata": {},
   "outputs": [],
   "source": [
    "iqtree3 -s concatenated.msa.alignment_mimic_100.fa -m LG+F+R4 -t ~/CoBi-course/full_dataset_concat/concatenated.msa.treefile --scfl 100 -pre concatenated.msa.alignment_mimic_100.fa.sCF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f739229-2d36-4c27-8180-b04851180d57",
   "metadata": {},
   "source": [
    "Here we will use the distribution of the sDF from these 100 simulated MSA to build the null distribution of sDF which lacks of `GTEE`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ec8614a4-92a0-454e-8540-3a3dc20d38c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "python ~/CoBi-course/func/multifunction.py quantifygtee ~/CoBi-course/Simulated_MSA_sCF/*.stat -o GTEE.svg"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b8554b-d917-420a-a59b-bea88e08a529",
   "metadata": {},
   "source": [
    "> Double click on the `GTEE.svg` file on the left side or open a `python` Console and type in the following commands and the `shift+enter` key to see the figure:\n",
    "> \n",
    "> ```python\n",
    "from IPython.display import SVG\n",
    "from pathlib import Path\n",
    "file_path = \"/home/jovyan/CoBi-GTEE/GTEE.svg\"\n",
    "display(SVG(Path(file_path)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
